rpcs3/Emu/Cell/lv2/sys_gpio.cpp

```diff
@@ -0,0 +1,38 @@
+#include "stdafx.h"
+#include "Emu/Memory/Memory.h"
+#include "Emu/System.h"
+#include "Emu/IdManager.h"
+
+#include "Emu/Cell/ErrorCodes.h"
+#include "sys_gpio.h"
+
+namespace vm { using namespace ps3; }
+
+logs::channel sys_gpio("sys_gpio");
+
+error_code sys_gpio_get(u64 device_id, vm::ptr<u64> value)
+{
+  sys_gpio.trace("gpio get (device_id = 0x%x ,value = *0x%x)",device_id ,value);
+  if (device_id != 1 && device_id != 2)
+  {
+    return CELL_ESRCH;
+  }
+
+  device_id--;
+  *value = gpio_entries[device_id];
+  return CELL_OK;
+}
+
+error_code sys_gpio_set(u64 device_id, u64 mask, u64 value)
+{
+  sys_gpio.trace("gpio set (device_id = 0x%x ,mask = 0x%x ,value = 0x%x)",device_id ,mask ,value);
+  if (device_id != 1 && device_id != 2)
+  {
+    return CELL_ESRCH;
+  }
+
+  device_id--;
+  gpio_entries[device_id] = value & 0xf & mask;
```

Comment:
Assuming that the mask indicates the bits to change and the value to indicate the new state this would be:
gpio_entries[device_id] = ((gpio_entries[device_id] & ~mask) | (value & mask)) & 0xF;

This allows setting and resetting of individual bits.

Reply:
only the firt 4 bits are used ,its mention in the sdk

Reply:
interesting ,if thats the case then it makes much more sence to pass a mask to the syscall. but It needs to be tested on the decr.

Reply:
@jjsat oddly its masking the value that set e.g `value & mask` ,read it on one of ps3 dev wiki linked tutorials